File: solutions/python/2020/day10.py
from common2020 import int_list, print_ans


def pt_1():
    diff1 = 0
    diff3 = 1
    current = 0
    target = max(joltage_list)

    while current < target:
        if(current + 1 in joltage_list):
            current += 1
            diff1 += 1
        elif(current + 2 in joltage_list):
            current += 2
        elif(current + 3 in joltage_list):
            current += 3
            diff3 += 1
        else:
            return -1

    return diff1 * diff3


def pt_2():
    sol = {0:1}
    for line in sorted(joltage_list):
        sol[line] = 0
        if line - 1 in sol:
            sol[line] += sol[line-1]
        if line - 2 in sol:
            sol[line] += sol[line-2]
        if line - 3 in sol:
            sol[line] += sol[line-3]
    
    return sol[max(joltage_list)]


def main():
    pt1 = pt_1()
    pt2 = pt_2()
    print_ans(pt1, pt2)


joltage_list = int_list(10)
main()

File: solutions/2020/day10.ts
import { input, print } from "common";
import { last } from "ramda";

function joltageList(): number[] {
  return input(2020, 10)
    .split("\n")
    .map((num) => parseInt(num))
    .sort((a, b) => a - b);
}

function pt1(jl: number[]): number {
  const target = last(jl);
  let diff1 = 0;
  let diff3 = 1;
  let current = 0;

  while (current < target) {
    if (jl.includes(current + 1)) {
      current += 1;
      diff1 += 1;
    } else if (jl.includes(current + 2)) {
      current += 2;
    } else if (jl.includes(current + 3)) {
      current += 3;
      diff3 += 1;
    } else {
      return -1;
    }
  }

  return diff1 * diff3;
}

function pt2(jl: number[]): number {
  const sol: Record<number, number> = { 0: 1 };
  for (const line of jl) {
    sol[line] = 0;
    if (sol[line - 1]) {
      sol[line] += sol[line - 1];
    }
    if (sol[line - 2]) {
      sol[line] += sol[line - 2];
    }
    if (sol[line - 3]) {
      sol[line] += sol[line - 3];
    }
  }

  return sol[last(jl)];
}

print(pt1(joltageList()), pt2(joltageList()));
